File: bindings/zig/src/safe_crypto.zig
// SPDX-License-Identifier: PMPL-1.0-or-later
// Copyright (c) 2026 Jonathan D.A. Jewell (hyperpolymath) <jonathan.jewell@open.ac.uk>
//
// Proven SafeCrypto - FFI bindings to libproven cryptographic operations.
// All computation is performed in verified Idris 2 code via libproven.

const std = @import("std");
const c = @cImport(@cInclude("proven.h"));

/// Error types for crypto operations.
pub const CryptoError = error{
    ProvenError,
};

/// Constant-time byte comparison (timing-attack safe) via libproven.
/// Returns false if lengths differ.
pub fn constantTimeCompare(a: []const u8, b: []const u8) bool {
    const result = c.proven_crypto_constant_time_eq(
        a.ptr,
        a.len,
        b.ptr,
        b.len,
    );
    return result.status == c.PROVEN_OK and result.value;
}

/// Fill buffer with cryptographically secure random bytes via libproven.
pub fn randomBytes(buffer: []u8) CryptoError!void {
    const status = c.proven_crypto_random_bytes(buffer.ptr, buffer.len);
    if (status != c.PROVEN_OK) return error.ProvenError;
}

test "constantTimeCompare" {
    try std.testing.expect(constantTimeCompare("secret", "secret"));
    try std.testing.expect(!constantTimeCompare("secret", "other!"));
    try std.testing.expect(!constantTimeCompare("short", "longer"));
}

test "randomBytes" {
    var buf: [16]u8 = undefined;
    try randomBytes(&buf);
}

File: bindings/rust/src/safe_crypto.rs
// SPDX-License-Identifier: PMPL-1.0-or-later
// Copyright (c) 2026 Jonathan D.A. Jewell (hyperpolymath) <jonathan.jewell@open.ac.uk>

//! Safe cryptographic primitives via libproven FFI.
//!
//! Provides constant-time comparison and cryptographic random byte generation.
//! All operations delegate to Idris 2 verified code.

use crate::core::{self, Error, Result};
use crate::ffi;

/// Safe cryptographic operations.
pub struct SafeCrypto;

impl SafeCrypto {
    /// Constant-time byte comparison (timing-attack resistant).
    ///
    /// Returns `true` if both byte slices are identical. The comparison
    /// takes the same amount of time regardless of where a mismatch occurs.
    pub fn constant_time_eq(a: &[u8], b: &[u8]) -> Result<bool> {
        // SAFETY: We pass valid pointers and lengths from Rust slices.
        // The FFI function only reads from both pointers within their lengths.
        let result = unsafe {
            ffi::proven_crypto_constant_time_eq(
                a.as_ptr(),
                a.len(),
                b.as_ptr(),
                b.len(),
            )
        };
        core::bool_result_to_result(result)
    }

    /// Fill a buffer with cryptographically secure random bytes.
    ///
    /// Uses the Idris 2 verified CSPRNG implementation.
    pub fn random_bytes(buf: &mut [u8]) -> Result<()> {
        // SAFETY: We pass a valid mutable pointer and length from a Rust slice.
        // The FFI function writes to the pointer within the given length.
        let status = unsafe {
            ffi::proven_crypto_random_bytes(buf.as_mut_ptr(), buf.len())
        };
        core::status_to_result(status)
    }

    /// Generate a vector of cryptographically secure random bytes.
    ///
    /// Convenience wrapper around `random_bytes` that allocates the buffer.
    pub fn random_vec(len: usize) -> Result<Vec<u8>> {
        let mut buf = vec![0u8; len];
        Self::random_bytes(&mut buf)?;
        Ok(buf)
    }
}
